### PR TITLE
If SQL file already scheduled for parsing, don't reprocess

### LINKED ISSUE #3589: [PP] IndexError in update_mssat_in_saved
I experienced this error after switching branches. The branch I switched to had three additional models and one additional analysis.

```
2021-07-16 18:34:23.889246 (MainThread): Partial parsing enabled: 0 files deleted, 4 files added, 6 files changed.
2021-07-16 18:34:23.893682 (MainThread): Partial parsing: added file: fishtown_internal_analytics://models/marts/dbt/dbt_experimental_parser_samples.sql
2021-07-16 18:34:23.893852 (MainThread): Partial parsing: added file: fishtown_internal_analytics://models/marts/dbt/dbt_experimental_parser_misses.sql
2021-07-16 18:34:23.893953 (MainThread): Partial parsing: added file: fishtown_internal_analytics://analysis/experimental_parser_missed_files.sql
2021-07-16 18:34:23.894057 (MainThread): Partial parsing: added file: fishtown_internal_analytics://models/staging/snowplow/raw/current/dbt/current_snowplow_experimental_parser.sql
2021-07-16 18:34:23.894222 (MainThread): Partial parsing: update schema file: fishtown_internal_analytics://models/staging/snowplow/schema.yml
2021-07-16 18:34:23.894419 (MainThread): Partial parsing: update schema file: fishtown_internal_analytics://models/marts/dbt/schema.yml
2021-07-16 18:34:23.894523 (MainThread): Partial parsing: updated file: fishtown_internal_analytics://models/staging/snowplow/raw/current/dbt/current_snowplow_load_all_timing.sql
2021-07-16 18:34:23.894736 (MainThread): Partial parsing: deleted exposure exposure.fishtown_internal_analytics.dbt_minor_version_adoption
2021-07-16 18:34:23.894890 (MainThread): Partial parsing: deleted exposure exposure.fishtown_internal_analytics.metrics_dashboard
2021-07-16 18:34:23.895036 (MainThread): Partial parsing: deleted exposure exposure.fishtown_internal_analytics.weekly_dbt_metrics
2021-07-16 18:34:23.895196 (MainThread): Partial parsing: updated file: fishtown_internal_analytics://models/marts/dbt/docs.md
2021-07-16 18:34:23.895365 (MainThread): Sending event: {'category': 'dbt', 'action': 'invocation', 'label': 'end', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x10e4a4340>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x10e4a4310>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x10e4a42b0>]}
2021-07-16 18:34:23.895583 (MainThread): Flushing usage events
2021-07-16 18:34:24.018905 (MainThread): Encountered an error:
2021-07-16 18:34:24.019179 (MainThread): list index out of range
2021-07-16 18:34:24.025747 (MainThread): Traceback (most recent call last):
  File "/usr/local/Cellar/dbt/0.20.0_1/libexec/lib/python3.8/site-packages/dbt/main.py", line 125, in main
    results, succeeded = handle_and_check(args)
  File "/usr/local/Cellar/dbt/0.20.0_1/libexec/lib/python3.8/site-packages/dbt/main.py", line 203, in handle_and_check
    task, res = run_from_args(parsed)
  File "/usr/local/Cellar/dbt/0.20.0_1/libexec/lib/python3.8/site-packages/dbt/main.py", line 256, in run_from_args
    results = task.run()
  File "/usr/local/Cellar/dbt/0.20.0_1/libexec/lib/python3.8/site-packages/dbt/task/runnable.py", line 406, in run
    self._runtime_initialize()
  File "/usr/local/Cellar/dbt/0.20.0_1/libexec/lib/python3.8/site-packages/dbt/task/runnable.py", line 122, in _runtime_initialize
    super()._runtime_initialize()
  File "/usr/local/Cellar/dbt/0.20.0_1/libexec/lib/python3.8/site-packages/dbt/task/runnable.py", line 79, in _runtime_initialize
    self.load_manifest()
  File "/usr/local/Cellar/dbt/0.20.0_1/libexec/lib/python3.8/site-packages/dbt/task/runnable.py", line 66, in load_manifest
    self.manifest = ManifestLoader.get_full_manifest(self.config)
  File "/usr/local/Cellar/dbt/0.20.0_1/libexec/lib/python3.8/site-packages/dbt/parser/manifest.py", line 170, in get_full_manifest
    manifest = loader.load()
  File "/usr/local/Cellar/dbt/0.20.0_1/libexec/lib/python3.8/site-packages/dbt/parser/manifest.py", line 212, in load
    project_parser_files = partial_parsing.get_parsing_files()
  File "/usr/local/Cellar/dbt/0.20.0_1/libexec/lib/python3.8/site-packages/dbt/parser/partial.py", line 131, in get_parsing_files
    self.update_in_saved(file_id)
  File "/usr/local/Cellar/dbt/0.20.0_1/libexec/lib/python3.8/site-packages/dbt/parser/partial.py", line 199, in update_in_saved
    self.update_mssat_in_saved(new_source_file, old_source_file)
  File "/usr/local/Cellar/dbt/0.20.0_1/libexec/lib/python3.8/site-packages/dbt/parser/partial.py", line 215, in update_mssat_in_saved
    unique_id = old_source_file.nodes[0]
IndexError: list index out of range
```

### PR DESCRIPTION
**Pull request by gshank**
resolves #3589

### Description

Skip re-processing SQL file if already scheduled.

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

**Comment by gshank** [comment]
It's not clear to me from the bug description how exactly this occurs. From the code it looks like we are attempting to process a changed SQL file that's already been processed. The change in this PR should be okay in any case, but I'm not sure how this happens so was not able to do a test. If someone can provide a test case, I'd appreciate it.

File: core/dbt/parser/partial.py [hunk 1]
```diff
@@ -212,6 +212,10 @@ def update_in_saved(self, file_id):
     def update_mssat_in_saved(self, new_source_file, old_source_file):
 
         # These files only have one node.
```
Comment (leahwicz) [inline core/dbt/parser/partial.py:214]: Should this comment be below the `if` statement with the original code it was with?

**Comment by gshank** [comment]
I suspect that what happened was that some models were deleted and models that referred to those models were scheduled for parsing, and then the code attempted to update them. I should be able to come up with a test case next week.

File: core/dbt/parser/partial.py (hunk 1 above)
Comment (gshank) [reply core/dbt/parser/partial.py]: Yes. Addressed in latest version.

**Review by nathaniel-may — APPROVED** [review]

File: core/dbt/parser/partial.py [hunk 2]
```diff
@@ -211,6 +223,9 @@ def update_in_saved(self, file_id):
     # Updated schema files should have been processed already.
     def update_mssat_in_saved(self, new_source_file, old_source_file):
 
+        if self.already_scheduled_for_parsing(old_source_file):
```
Comment (iknox-fa) [inline core/dbt/parser/partial.py:226]: Could we dry these checks up a little by putting a single check in `update_in_saved`?  AFAICT that's the only place these functions are called.

**Review by iknox-fa — CHANGES_REQUESTED** [review]
[One small change.](https://github.com/dbt-labs/dbt/pull/3620/files#r676876508)

File: core/dbt/parser/partial.py (hunk 2 above)
Comment (gshank) [reply core/dbt/parser/partial.py:226]: That would work, but I'm hoping to clean up the factoring of this code in the future and reuse these functions, so I'm going to leave them as separate calls for now.

**Pull request merged**